python/ray/_private/job_manager/job_manager.py

```diff
@@ -152,21 +156,66 @@ def run(self, cmd: str):
             stdout_path, stderr_path = self._log_client.get_log_file_paths(
                 self._job_id)
 
-            exit_code = exec_cmd_logs_to_file(cmd, stdout_path, stderr_path)
+            exit_code = await self._exec_cmd(cmd, stdout_path, stderr_path)
+        except asyncio.CancelledError:
+            logger.debug("Driver command coroutine canceled.")
+        except Exception as e:
+            logger.debug(e)
         finally:
+            logger.debug(
+                f"Driver command coroutine returned, exit code: {exit_code}")
             # 3) Once command finishes, update status to SUCCEEDED or FAILED.
-            if exit_code == 0:
-                self._status = JobStatus.SUCCEEDED
-            else:
-                self._status = JobStatus.FAILED
-            self._status_client.put_status(self._job_id, self.get_status())
-            ray.actor.exit_actor()
+            # No action if command is stopped by user
+            cur_status = self.get_status()
+            if cur_status != JobStatus.STOPPED and exit_code is not None:
+                # Update terminal status based on subprocess return code.
+                if exit_code == 0:
+                    self._status_client.put_status(self._job_id,
+                                                   JobStatus.SUCCEEDED)
+                else:
+                    self._status_client.put_status(self._job_id,
+                                                   JobStatus.FAILED)
+
+            if exit_code is not None:
+                # Exit code as None indicates driver command is terminated
+                # by calling stop() which cancels task coroutine. In this case
+                # we want to ensure stop() finishes with returned status and
+                # offload supervisor actor exit at job manager level to avoid
+                # race condition where supervisor actor died before
+                # job_supervisor_actor.stop.remote() returns.
+                logger.debug(
+                    "Gracefully self-exiting job supervisor actor after "
+                    "driver command finished execution.")
+                ray.actor.exit_actor()
```

Comment:
calling ray.actor.exit_actor causes errors to be propogated (when I was testing). If you look at the code in the product, we pass an actor handle to get around this.